src/mesh/mod.rs

```diff
@@ -796,6 +822,35 @@ fn remove_orphan_vertices_and_normals(
     (faces_renumbered, vertices_reduced, normals_reduced)
 }
 
+pub fn calculate_smooth_normals_from_components(
+    vertices: &[Point3<f32>],
+    faces: &[Face],
+    vertex_to_face_topology: &[SmallVec<[u32; topology::MAX_INLINE_NEIGHBOR_COUNT]>],
+) -> Vec<Vector3<f32>> {
+    let vertex_count = vertices.len();
+
+    let mut normals: Vec<Vector3<f32>> = Vec::with_capacity(vertex_count);
+
+    for shared_face_indices in vertex_to_face_topology {
+        let shared_faces_count_f32 = shared_face_indices.len() as f32;
+        let mut normal: Vector3<f32> = Vector3::zeros();
+        for face_index in shared_face_indices {
+            match faces[cast_usize(*face_index)] {
+                Face::Triangle(face) => {
+                    normal += geometry::compute_triangle_normal(
+                        &vertices[cast_usize(face.vertices.0)],
+                        &vertices[cast_usize(face.vertices.1)],
+                        &vertices[cast_usize(face.vertices.2)],
+                    ) / shared_faces_count_f32;
```

Comment:
This will end up being normalized, right?

Reply:
Now this is the crucial question! I've been thinking the same when I realized we never guarantee the mesh normals are normalized! I believe we should normalize all the normals and I think the best place would be the mesh constructor. 
The reason for normalizing are comparison methods and welding (see the other PR).

Reply:
i decided to normalize the normals in mesh constructors